File: Daily Challenge/February 2025/Remove All Occurrences of a Substring.cs
public class Solution {
  public string RemoveOccurrences(string s, string part) {
    int sLength = s.Length;
    int partLength = part.Length;
    Stack<char> characterStack = new Stack<char>();

    foreach (char x in s) {
      characterStack.Push(x);

      if (characterStack.Count >= partLength && StackContainsPart(characterStack, ref part)) {
        for (int i = 0; i < partLength; ++i) {
          characterStack.Pop();
        }
      }
    }

    string result = "";
    while (characterStack.Count > 0) {
      result = characterStack.Pop() + result;
    }

    return result;
  }

  private bool StackContainsPart(Stack<char> characterStack, ref string part) {
    Stack<char> tempStack = new Stack<char>();
    
    for (int i = part.Length - 1; i >= 0; --i) {
      if (characterStack.Peek() != part[i]) {
        while (tempStack.Count > 0) {
          characterStack.Push(tempStack.Pop());
        }

        return false;
      }

      tempStack.Push(characterStack.Pop());
    }

    while (tempStack.Count > 0) {
      characterStack.Push(tempStack.Pop());
    }

    return true;
  }
}

File: Daily Challenge/February 2025/Remove All Occurrences of a Substring.cpp
class Solution {
public:
  string removeOccurrences(string s, string part) {
    int sLength = s.size();
    int partLength = part.size();
    stack<char> charStack;

    for (char x : s) {
      charStack.push(x);

      if (charStack.size() >= partLength and stackContainsPart(charStack, part)) {
        for (int i = 0; i < partLength; ++i) {
          charStack.pop();
        }
      }
    }

    string result = "";
    while (!charStack.empty()) {
      result = charStack.top() + result;
      charStack.pop();
    }

    return result;
  }

private:
  bool stackContainsPart(stack<char> charStack, string& part) {
    int partLength = part.size();

    for (int i = partLength - 1; i >= 0; --i) {
      if (charStack.top() != part[i]) return false;
      charStack.pop();
    }

    return true;
  }
};